File: python/src/a2a_utils/tasks/json_task_store.py
"""JSON file-based task store implementing the A2A SDK TaskStore ABC."""

import asyncio
import json
from pathlib import Path

from a2a.server.tasks import TaskStore
from a2a.types import Task
from loguru import logger


class JSONTaskStore(TaskStore):
    """Persists Task objects as individual JSON files.

    Each task is stored at ``{storage_dir}/{task_id}.json`` using
    Pydantic's ``model_dump(mode="json")`` for serialization.
    """

    def __init__(self, storage_dir: Path) -> None:
        self._storage_dir = storage_dir
        self._storage_dir.mkdir(parents=True, exist_ok=True)
        self._lock = asyncio.Lock()

    async def save(self, task: Task) -> None:  # type: ignore[override]
        """Save a task to disk.

        Args:
            task: The Task object to persist.
        """
        async with self._lock:
            file_path = self._storage_dir / f"{task.id}.json"
            data = task.model_dump(mode="json")
            file_path.write_text(json.dumps(data, indent=2))
            logger.debug(f"Saved task {task.id} to {file_path}")

    async def get(self, task_id: str) -> Task | None:  # type: ignore[override]
        """Load a task from disk.

        Args:
            task_id: The task ID to look up.

        Returns:
            Task if found, None otherwise.
        """
        file_path = self._storage_dir / f"{task_id}.json"
        if not file_path.exists():
            return None

        try:
            data = json.loads(file_path.read_text())
            return Task.model_validate(data)
        except Exception as e:
            logger.error(f"Error loading task {task_id}: {e}")
            return None

    async def delete(self, task_id: str) -> None:  # type: ignore[override]
        """Delete a task from disk.

        Args:
            task_id: The task ID to delete.
        """
        async with self._lock:
            file_path = self._storage_dir / f"{task_id}.json"
            if file_path.exists():
                file_path.unlink()
                logger.debug(f"Deleted task {task_id}")

File: javascript/src/tasks/json-task-store.ts
// SPDX-FileCopyrightText: 2025-present A2A Net <hello@a2anet.com>
//
// SPDX-License-Identifier: Apache-2.0

/**
 * JSON file-based task store implementing the A2A SDK TaskStore interface.
 */

import * as fs from "node:fs/promises";
import * as path from "node:path";
import type { Task } from "@a2a-js/sdk";
import type { TaskStore } from "@a2a-js/sdk/server";

/**
 * Persists Task objects as individual JSON files.
 *
 * Each task is stored at ``{storageDir}/{taskId}.json``.
 */
export class JSONTaskStore implements TaskStore {
    readonly _storageDir: string;

    constructor(storageDir: string) {
        this._storageDir = storageDir;
        // Ensure storage dir exists
        fs.mkdir(storageDir, { recursive: true }).catch(() => {});
    }

    /**
     * Save a task to disk.
     *
     * @param task - The Task object to persist.
     */
    async save(task: Task): Promise<void> {
        await fs.mkdir(this._storageDir, { recursive: true });
        const filePath = path.join(this._storageDir, `${task.id}.json`);
        const data = JSON.stringify(task, null, 2);
        await fs.writeFile(filePath, data);
        console.debug(`Saved task ${task.id} to ${filePath}`);
    }

    /**
     * Load a task from disk.
     *
     * @param taskId - The task ID to look up.
     *
     * @returns Task if found, undefined otherwise.
     */
    async load(taskId: string): Promise<Task | undefined> {
        const filePath = path.join(this._storageDir, `${taskId}.json`);
        try {
            const text = await fs.readFile(filePath, "utf-8");
            return JSON.parse(text) as Task;
        } catch {
            return undefined;
        }
    }

    /**
     * Delete a task from disk.
     *
     * @param taskId - The task ID to delete.
     */
    async delete(taskId: string): Promise<void> {
        const filePath = path.join(this._storageDir, `${taskId}.json`);
        try {
            await fs.unlink(filePath);
            console.debug(`Deleted task ${taskId}`);
        } catch {
            // Ignore if doesn't exist
        }
    }
}
